pkg/objstore/bos/bos.go

```diff
@@ -0,0 +1,341 @@
+package bos
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/baidu/baiducloud-sdk-go/bce"
+	"github.com/baidu/baiducloud-sdk-go/bos"
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const dirDelim = "/"
+
+// Bucket implements the store.Bucket interface against bos-compatible(Baidu Object Storage) APIs.
+type Bucket struct {
+	logger log.Logger
+	client *bos.Client
+	name   string
+}
+
+// Config encapsulates the necessary config values to instantiate an bos client.
+type Config struct {
+	Bucket          string `yaml:"bucket"`
+	Region          string `yaml:"region"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	SecretAccessKey string `yaml:"secret_access_key"`
+}
+
+func (conf *Config) validate() error {
+	if conf.Bucket == "" ||
+		conf.Region == "" ||
+		conf.AccessKeyID == "" ||
+		conf.SecretAccessKey == "" {
+		return errors.New("insufficient bos configuration information")
+	}
+	return nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	var config *Config
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, errors.Wrap(err, "parsing bos configuration")
+	}
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate bos configuration")
+	}
+
+	client := bos.NewClient(&bce.Config{
+		Region: config.Region,
+		Credentials: &bce.Credentials{
+			AccessKeyID:     config.AccessKeyID,
+			SecretAccessKey: config.SecretAccessKey,
+		},
+	})
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		name:   config.Bucket,
+	}
+	return bkt, nil
+}
+
+// Name returns the bucket name for the provider.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	if err := b.client.DeleteObject(b.name, name, nil); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	if _, err := b.client.PutObject(b.name, name, r, nil, nil); err != nil {
```

Comment:
I saw the official documents that the sample upload interface could not upload the file large than 5GB, sometimes the tsdb files (index) would exceed 5GB. Could you double check this interface would not be impact by the limitation?